dag/dag.go

```diff
@@ -0,0 +1,239 @@
+// Copyright 2021 Mineiros GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dag
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/madlambda/spells/errutil"
+)
+
+type (
+	// ID of nodes
+	ID string
+
+	// DAG is a Directed-Acyclic Graph
+	DAG struct {
+		dag    map[ID][]ID
+		values map[ID]interface{}
+		cycles map[ID]bool
+
+		validated bool
+	}
+)
+
+const (
+	ErrDuplicateNode errutil.Error = "duplicate node"
+	ErrNodeNotFound  errutil.Error = "node not found"
+	ErrCycleDetected errutil.Error = "cycle detected"
+)
+
+// New creates a new empty Directed-Acyclic-Graph.
+func New() *DAG {
+	return &DAG{
+		dag:    make(map[ID][]ID),
+		values: make(map[ID]interface{}),
+	}
+}
+
+// AddNode adds a new node to the dag. The lists of before and after
+// defines its edge nodes.
+func (d *DAG) AddNode(id ID, value interface{}, before, after []ID) error {
+	if _, ok := d.values[id]; ok {
+		return errutil.Chain(
+			ErrDuplicateNode,
+			fmt.Errorf("adding node id %q", id),
+		)
+	}
+
+	for _, bid := range before {
+		if _, ok := d.dag[bid]; !ok {
+			d.dag[bid] = []ID{}
```

Comment:
the value is not important here, but the key. We're initializing the map key `bid` because this key was seen in the `before` of another node. This is needed to make possible iterative validations, or better saying we can validate the dag at any point in time, and it validates the current state of the dag, so if we don't initialize the key when it's seen, the validator will not be able to reach those keys (this is used by `plan graph` because it should be able to print a graph with cycles). So in this case it could be just `d.dag[bid] = nil` but this is not nice because a non-existent key also returns nil...